### PR TITLE
Predefined constants and annotation keywords

### PR DESCRIPTION
**Pull request by Edih112**
**Overview**
* Added predefined constants in Antimony to the grammar along with error checking and hover information. 
Aims to address issue:  https://github.com/sys-bio/AntimonyEditor/issues/39

* Also updated the known annotation keywords within the grammar file based on the current keywords found in the following:
Model qualifiers:  https://github.com/sys-bio/antimony/blob/develop/src/annotated.cpp#L196
Biological qualifiers:  https://github.com/sys-bio/antimony/blob/develop/src/annotated.cpp#L231

**Main Changes**
* `ModelSemanticChecker.ts`: added hover information for predefined constants.
*  `SemanticErrors.ts`: updated the incompatible types and overriding value errors messages to account for built-in constant cases.
* `SemanticVisitor.ts`:  updated logic to only complain about lack of initialization for non built-in constants.
* `SymbolTableClasses.ts`: updated the global symbol table to always have the built-in constants defined.
*  `SymboleTableVisitor.ts`:  added error checking for reassigning values to built-in constants or trying to define another variable using the same id as a built-in  constant.

The built-in constant updates can also be seen in the following:
image
image
image

* `AntimonyGrammar.g4`: added new annotation keywords.
    * These annotation keywords include the following:
       * `is`
       * `model_source`
       * `biological_system`
       * `publication`
       * `isDerivedFrom`
       * `hasInstance`
       * `instance`

**Tests**
Updated `App.test` to include tests for cases shown above.

### CODE REVIEW

File: src/language-handler/SemanticVisitor.ts [hunk 1]
```diff
@@ -139,11 +139,14 @@ export class SemanticVisitor extends ErrorVisitor implements AntimonyGrammarVisi
       const varInfo = currST.getVar(varName);
       if (this.currNameAndScope?.scope !== 'function' && varInfo && varInfo.initSrcRange === undefined) {
         if (varInfo.type === varTypes.Parameter) {
-          // error, needs initialized value
-          //Parameter 'k' missing value assignment
-          const errorMessage: string = unitializedParameterError(varName);
-          const errorUnderline: ErrorUnderline = this.getErrorUnderline(idSrcRange, errorMessage, true);
-          this.addError(errorUnderline);
+          // need to check if this is a predefined-constatnt which does not require initializatino
```
Comment (luciansmith) [inline src/language-handler/SemanticVisitor.ts:142]: typo; 'initializatino'

File: src/testing/App.test.tsx [hunk 2]
```diff
@@ -155,6 +155,124 @@ jest.mock('monaco-editor', () => ({
...
+                "endLineNumber": 3,
+                "message": "Unabled to set type to 'species' because it is a predefined constant",
+                "severity": 8,
+                "startColumn": 9,
+                "startLineNumber": 3,
+            },
+            {
+                "endColumn": 0,
+                "endLineNumber": 0,
+                "message": "Value assignment to 'NaN' is being overridden by a later assignment on line 3:9",
```
Comment (luciansmith) [inline src/testing/App.test.tsx:205]: This is a weird error message for 'NaN'.  You can't *ever* assign a value to it, so the 'later assignment' also wouldn't work.

File: src/testing/App.test.tsx [hunk 3]
```diff
@@ -155,6 +155,124 @@ jest.mock('monaco-editor', () => ({
...
+                "endLineNumber": 2,
+                "message": "Unabled to set type to 'species' because it is a predefined constant",
+                "severity": 8,
+                "startColumn": 9,
+                "startLineNumber": 2,
+            },
+            {
+                "endColumn": 0,
+                "endLineNumber": 0,
+                "message": "Value assignment to 'inf' is being overridden by a later assignment on line 2:9",
```
Comment (luciansmith) [inline src/testing/App.test.tsx:252]: (same comment here as for NaN)

File: src/testing/testAntFiles/predefConstantTypeAssignError.ant [hunk 4]
```diff
@@ -0,0 +1,4 @@
+species pi;
+species inf = 1;
+compartment avogadro;
+const species Nan = 0;
```
Comment (luciansmith) [inline src/testing/testAntFiles/predefConstantTypeAssignError.ant:4]: The line should be fine, right?  It shouldn't trigger an error.

Similarly 'species Pi' should also not cause an error.

File: src/testing/App.test.tsx (hunk 3 above)
Comment (Edih112) [reply src/testing/App.test.tsx:252]: Will fix in next commit

File: src/testing/App.test.tsx (hunk 2 above)
Comment (Edih112) [reply src/testing/App.test.tsx:205]: Will update to not report this in next commit

File: src/language-handler/SemanticVisitor.ts (hunk 1 above)
Comment (Edih112) [reply src/language-handler/SemanticVisitor.ts:142]: Will change in next commit.

File: src/testing/testAntFiles/predefConstantTypeAssignError.ant (hunk 4 above)
Comment (Edih112) [reply src/testing/testAntFiles/predefConstantTypeAssignError.ant:4]: Exactly! Nan does not cause an error here, I will add `species Pi `to this test case as well!

**Review by luciansmith — APPROVED** [review]

File: src/language-handler/SemanticErrors.ts [hunk 5]
```diff
@@ -10,10 +10,20 @@ import { Variable } from "./Variable";
  * @returns the error message string
  */
 export function incompatibleTypesError(newType: varTypes, oldVar: Variable): string {
-  const errorMessage: string = "Unable to set the type to '" + newType +
-                               "' because it is already set to be the incompatible type '"+ oldVar.type +
-                               "' on line " + 
-                               oldVar.idSrcRange.start.toString();
+  let errorMessage: string
+  if (oldVar.type === varTypes.PredefConstant) {
+    errorMessage = "Unabled to set type to '" + newType + "' because it is a predefined constant";
```
Comment (kevincnguyen) [inline src/language-handler/SemanticErrors.ts:15]: Typo; Should be 'Unable' instead of 'Unabled'

File: src/language-handler/SemanticErrors.ts [hunk 6]
```diff
@@ -10,10 +10,20 @@ import { Variable } from "./Variable";
...
+    errorMessage= "Unable to set the type to '" + newType +
+                "' because it is already set to be the incompatible type '"+ oldVar.type +
+                "' on line " + 
+                oldVar.idSrcRange.start.toString();
+  }
+  return errorMessage;
+}
+
+export function predefConstantValueAssignmentError(id: string) {
+  const errorMessage= "Cannot assign value to " + id + " as it is a predefined constant";
```
Comment (kevincnguyen) [inline src/language-handler/SemanticErrors.ts:26]: For consistency, I think we should add '' around `id` as all the other error messages have '' for the predefined constant.

**Review by kevincnguyen — APPROVED** [review]
A few nitpicks, but LGTM!

File: src/language-handler/SemanticErrors.ts (hunk 5 above)
Comment (Edih112) [reply src/language-handler/SemanticErrors.ts:15]: Whoops, good catch! Will fix this in next commit.

File: src/language-handler/SemanticErrors.ts (hunk 6 above)
Comment (Edih112) [reply src/language-handler/SemanticErrors.ts]: Fixed.

File: src/language-handler/SemanticErrors.ts (hunk 5 above)
Comment (Edih112) [reply src/language-handler/SemanticErrors.ts]: Fixed.

File: src/language-handler/SemanticVisitor.ts [hunk 7]
```diff
@@ -139,11 +139,14 @@ export class SemanticVisitor extends ErrorVisitor implements AntimonyGrammarVisi
       const varInfo = currST.getVar(varName);
       if (this.currNameAndScope?.scope !== 'function' && varInfo && varInfo.initSrcRange === undefined) {
         if (varInfo.type === varTypes.Parameter) {
-          // error, needs initialized value
-          //Parameter 'k' missing value assignment
-          const errorMessage: string = unitializedParameterError(varName);
-          const errorUnderline: ErrorUnderline = this.getErrorUnderline(idSrcRange, errorMessage, true);
-          this.addError(errorUnderline);
+          // need to check if this is a predefined-constatnt which does not require initialization
```
Comment (konankisa) [inline src/language-handler/SemanticVisitor.ts:142]: Typo with constant

**Review by konankisa — APPROVED** [review]
Small typo, otherwise looks good to me!

**Pull request merged**